src/charm.py

```diff
@@ -107,6 +115,92 @@ def __init__(self, *args):
+
+        if not self._database_is_available():
+            return WaitingStatus("Waiting for the amf database to be available")
+
+        if not self._get_database_info():
+            return WaitingStatus("Waiting for AMF database info to be available")
+
+        if not self._nrf_requires.nrf_url:
+            return WaitingStatus("Waiting for NRF data to be available")
+
+        return None
+
+    def _check_container_availability(self) -> Optional[StatusBase]:
+        """Evaluate and return the unit's current status regarding with container/pod.
+
+        Returns:
+            StatusBase: MaintenanceStatus/BlockedStatus/WaitingStatus
+            None: If none of the conditionals match
+
+        """
+        if not self._amf_container.exists(path=CONFIG_DIR_PATH):
+            return WaitingStatus("Waiting for storage to be attached")
+
+        if not _get_pod_ip():
+            return WaitingStatus("Waiting for pod IP address to be available")
+
+        return None
+
+    def _is_unit_in_non_active_status(self) -> Optional[StatusBase]:
+        """Evaluate and return the unit's current status, or None if it should be active.
+
+        Returns:
+            StatusBase: MaintenanceStatus/BlockedStatus/WaitingStatus
+            None: If none of the conditionals match
+
+        """
+        if not self.unit.is_leader():
+            # NOTE: In cases where leader status is lost before the charm is
+            # finished processing all teardown events, this prevents teardown
+            # event code from running. Luckily, for this charm, none of the
+            # teardown code is necessary to preform if we're removing the
+            # charm.
+            return BlockedStatus("Scaling is not implemented for this charm")
+
+        if not self._amf_container.can_connect():
+            return MaintenanceStatus("Waiting for service to start")
+
+        if invalid_configs := self._get_invalid_configs():
+            return BlockedStatus(f"The following configurations are not valid: {invalid_configs}")
+
+        if relations_issues := self._check_relations():
+            return relations_issues
+
+        if container_issues := self._check_container_availability():
+            return container_issues
+        try:
+            self._set_n2_information()
+        except ValueError:
+            return BlockedStatus("Waiting for MetalLB to be enabled")
+
+        return None
+
+    def _on_collect_unit_status(self, event: CollectStatusEvent):
+        """Check the unit status and set to Unit when CollectStatusEvent is fired.
+
+        Args:
+            event: CollectStatusEvent
+        """
+        if status := self._is_unit_in_non_active_status():
+            event.add_status(status)
+        else:
+            if self._csr_is_stored() and not self._get_current_provider_certificate():
+                event.add_status(WaitingStatus("Waiting for certificates to be stored"))
+            event.add_status(ActiveStatus())
```

Comment:
Won't this always set `ActiveStatus()`?

Reply:
If there is any event which has higher priority than ActiveStatus, status set to that one.

```
If any statuses were added by the event handler using [add_status()](https://ops.readthedocs.io/en/latest/#ops.CollectStatusEvent.add_status), the framework will choose the highest-priority status and set that as the status (application status for collect_app_status, or unit status for collect_unit_status).

The order of priorities is as follows, from highest to lowest:
    error
    blocked
    maintenance
    waiting
    active
    unknown

If there are multiple statuses with the same priority, the first one added wins (and if an event is observed multiple times, the handlers are called in the order they were observed).
```

Reply:
In any cases, I arranged this method having the same logic but more explicit.